doc/reference/trd-spi.md

```diff
@@ -0,0 +1,359 @@
+`Some`. If `read_buffer` is `None` it does not affect the length of the operation.
+
+If the call to `read_write_bytes` returns `Ok(())`, the implementation MUST issue
+a callback to the `SpiControllerClient` when it completes. If the call returns
+an `Err`, the implementation MUST NOT issue a callback, except if the `ErrorCode` is
+`BUSY`. In this case, the implementation issues a callback for the outstanding
+operation but does not issue a callback for the failed one. If it returns `Err`,
+the implementation MUST return the buffers passed in the call. Valid `ErrorCode`
+values for an `Err` result are:
+  - BUSY: the SPI is busy with another call to `read_write_bytes` and so cannot
+    complete the request.
+  - OFF: the SPI is off and cannot accept a request.
+  - INVAL: the length value is 0, or one of the buffers passed has length 0.
+  - RESERVE: there is no client for a callback.
+  - FAIL: some other failure condition.
+
+The `set_client` method sets which callback to invoke when a `read_write_bytes`
+call completes. The `read_write_done` callback MUST return the buffers passed
+in the call to `read_write_bytes`. The `len` argument is the number of bytes
+read/written: it may differ from the `len` argument in `read_write_bytes` if
+one of the buffers was smaller than the call argument. The `status` argument
+indicates whether the SPI operation completed successfully. It may return any
+of the `ErrorCode` values that can be returned by `read_write_bytes`: these
+represent asynchronous errors (e.g., due to queueing).
+
+The `ControllerDevice` trait combines `Controller` and `Configure` traits.
+It provides the abstraction of being able to read/write to the bus and
+adjust its configuration.
+
+```rust
+pub trait ControllerDevice<'a>: Controller<'a> + Configure<'a> {}
+```
+
+4 `ChipSelect` and `ControllerBus`
+===============================
+
+The `ChipSelect` trait allows a client to change which chip select is
+active on the SPI bus. Because different SPI hardware can provide different
+numbers of chip selects, the actual chip select value is an associated
+type. This associated type is typically an `enum` so a chip implementation
+can statically verify that clients pass only valid chip select values.
+
+```rust
+pub trait ChipSelect {
+  type Value: Copy;
+  fn set_chip_select(&self, cs: Self::Value) -> Result<(), ErrorCode>;
+  fn get_chip_select(&self) -> Self::Value);
+}
+```
+
+The `ControllerBus` trait combines `ControllerDevice` and `ChipSelect` to
+provide the full abstraction of a SPI bus. It is the trait that chip
+SPI implementations provide. In addition to `ControllerDevice` and
+`ChipSelect`, `ControllerBus` includes an `init` method. This `init`
+method initializes the hardware to be a SPI controller and is typically
+called at boot.
+
+```rust
+pub trait ControllerBus<'a>: ControllerDevice<'a> + ChipSelect {
+  fn init(&self) -> Result<(), ErrorCode>;
```

Comment:
I know we had this discussion just last Friday, but it might make sense to be pedantic about it if we're formalizing it now: wouldn't we want to rather have this be `enable` and `disable`, such that one can switch between modes of underlying hardware. That's commonly supported and so I think we should expose it through this API for the advanced users.

The logic would be that a shared HW-device would need to have `spi::ControllerBus::disable` be called before something like `uart::ControllerBus::enable` can be called.

For HW which doesn't support deinitialization, we could just return an error on `disable`.

I'd like to avoid being able to "convert" a shared-HW device to a specific mode implicitly by calling `init`, without the need to uninitialize it explicitly first.

Reply:
I'd argue that the case you're describing would be neither of these APIs: it would be an API which supports moving between controller and peripheral mode. `init` is intended to be a one-time operation that's done at boot. These traits are intended for a SPI bus being in dedicated controller mode or dedicated peripheral mode. The fact that there's a syscall driver capsule that uses both is a bit of a historical artifact (and as far as I can tell, almost never used).

Reply:
Okay, that's one way of looking at the problem. I was hoping if we were to have an `enable`/`disable` API we could have our cake and eat it too: it could allow us to both support switching device modes as in the USART example, as well as selectively power down components when they are not used. It'd both just be a `disable` away.

Reply:
You want to separate enable/disable from power; some components can have significant power-up times (generally for things like oscillators to settle), such that you want to distinguish power state transitions from functional capabilities.